prqlc/prql-compiler/src/semantic/resolver/types.rs

```diff
@@ -72,22 +72,9 @@ impl Resolver<'_> {
                 let mut intersection = None;
                 for item in items {
                     let item_ty = Resolver::infer_type(item)?;
-
-                    if let Some(item_ty) = item_ty {
-                        if let Some(intersection) = &intersection {
-                            if intersection != &item_ty {
-                                // TODO: compute type intersection instead
-                                return Ok(None);
-                            }
-                        } else {
-                            intersection = Some(item_ty);
-                        }
-                    }
+                    intersection = maybe_type_intersection(intersection, item_ty);
```

Comment:
Wouldn't we generally call this a Union?

Reply:
That was my immediate impression as well and I came to make the same comment.

Reply:
This call will create a union, but not in all cases.

If you have `intersection(int, int)`, then this is just `int`.

If you have `intersection({a = int}, {a = bool})`, then this is `{a = int || bool}`.

Reply:
Aren't those unions? At least when I'm thinking about sets? :)

Reply:
`int` is not an union? it's a primitive set.

`{a = int || bool}` is not an union? it's a type of tuples that contain a union.

Reply:
I thought the definition of types that we liked was that *types* are the union of all the possible values that instances of that type can take?

Another indicator is the `||` / or operator which is a logical disjunction which goes hand in hand with unions, e.g. see [Logical Disjunction > Set Theory](https://en.wikipedia.org/wiki/Logical_disjunction#:~:text=%5D%5B6%5D-,Set%20theory,-%5Bedit%5D)

Reply:
Ok, you are now talking about terminology here. `int` is a set of all integer numbers, which is technically a union of all integers, i.e. `0 || 1 || -1 || 2 || -2, ...`.

For this specific union, we have a "primitive" set in PRQL, called `int`. So `int` is not a union in PRQL type system.

I think we are not understanding each other here. Can both of you formulate your questions very verbosely?

Reply:
I'm not worried about the `int` part, it's the `int || bool` part. My understanding is that `int = {0, 1, -1, 2, -2, ...}` and `bool = {true, false}` and `int || bool = {true, false} U {0, 1, -1, 2, -2, ...} = {true, false, 0, 1, -1, 2, -2, ...}`.

Hence it seems natural to refer to these as unions.

Reply:
Of course, `int || bool` is a union.

Are you concerned by the naming of functions here?

Reply:
Oh, only now I see what you mean - it's not the naming. We should be using a union of types of the items, not their intersection.